drivers/vulkan/vulkan_context.cpp

```diff
@@ -942,10 +907,10 @@ Error VulkanContext::_create_instance() {
 		/*sType*/ VK_STRUCTURE_TYPE_APPLICATION_INFO,
 		/*pNext*/ nullptr,
 		/*pApplicationName*/ cs.get_data(),
-		/*applicationVersion*/ 0,
+		/*applicationVersion*/ 0, // It would be really nice if we store a version number in project settings, say "application/config/version"
 		/*pEngineName*/ VERSION_NAME,
 		/*engineVersion*/ VK_MAKE_VERSION(VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH),
-		/*apiVersion*/ VK_MAKE_VERSION(vulkan_major, vulkan_minor, 0)
+		/*apiVersion*/ VK_API_VERSION_1_2 // We're targetting Vulkan 1.2, we can support older versions but we don't use features for newer versions. Note that setting this is purely used by validation layers.
```

Comment:
We can only set `apiVersion` to `VK_API_VERSION_1_2` when `vkEnumerateInstanceVersion` returns a version >= `VK_API_VERSION_1_1`. On a Vulkan 1.0 implementation this needs to be `VK_API_VERSION_1_0`. https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkApplicationInfo.html#_description

Now that `apiVersion` can either be `VK_API_VERSION_1_2` or `VK_API_VERSION_1_0`, we actually do need to cap `device_api_version` because `apiVersion` is the highest API version the application can use.

Reply:
Ok that seemed to contradict what @RandomShaper said, I'm happy to put back the code that checks instance version and cap this by the instance version. 

We can't cap by device version because we don't learn about the device version until after the instance is created.

That is also how I understood apiVersion, where we specify what version the API is build for and that this value is only used by the validation layer to warn us when we're calling an API that is not supported.

Reply:
I wasn't aware of the Vulkan 1.0 exception. Under Vulkan 1.0, you can't set `apiVersion` to anything but `API_VERSION_1_0`. On Vulkan 1.1+, we're fine with `API_VERSION_1_2`.

Reply:
@sakrel @RandomShaper have a look at my latest changes and see if this is better?

Reply:
Looks great! Still, I think `device_api_version` should be capped. The Vulkan spec says: 
> Device-level functionality or behavior added by a [new core version](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#versions) of the API must not be used unless it is supported by the device as determined by [VkPhysicalDeviceProperties](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkPhysicalDeviceProperties)::apiVersion and the specified version of [VkApplicationInfo](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkApplicationInfo)::apiVersion.

https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#fundamentals-validusage-versions

So, if `application_api_version` is 1.0 because we are on a Vulkan 1.0 instance, we can't use Vulkan 1.1 device features even when `device_api_version` is 1.1 or higher. Therefore, I think `device_api_version` should be set to `min(gpu_props.apiVersion, application_api_version)`.

Reply:
Ok, that contradicts everything that we've said before, that was how I had it working originally ?!?!

I'm soo confused now

Reply:
That's my understanding, too. We ask Vulkan for devices supporting 1.2 at least. If we get a 1.3 device, no problem because of backwards compatibility.

Reply:
But AFAIK we enable the use of 1.1 and 1.2 features conditionally. So we shouldn't mandate 1.2 support at this stage I suppose?

For the record, I would be fine with outright dropping Vulkan 1.0 support - the drivers from that era are too buggy to bother IMO. At least for now.

Reply:
Quoting the spec about `VkApplicationInfo`:
> `apiVersion` must be the highest version of Vulkan that the application is designed to use, encoded as described in [Version Numbers](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#extendingvulkan-coreversions-versionnumbers). [...]

> Vulkan 1.0 implementations were required to return `VK_ERROR_INCOMPATIBLE_DRIVER` if `apiVersion` was larger than 1.0. Implementations that support Vulkan 1.1 or later must not return `VK_ERROR_INCOMPATIBLE_DRIVER` for any value of `apiVersion`.

I believe the current code in this PR correctly accounts for the above. `VkApplicationInfo::apiVersion` is more informative than prescriptive.

Reply:
> That's my understanding, too. We ask Vulkan for devices supporting 1.2 at least. If we get a 1.3 device, no problem because of backwards compatibility.

Yes, there's no requirement to cap the device version on a Vulkan 1.1+ instance. I just figured that it would be safer to cap the version anyway. This way if somebody adds `if (device_api_version >= VK_API_VERSION_1_3)` checks, they would be forced to first raise `VkApplicationInfo::apiVersion` to get the check to pass. On second thought, this is something validation layers (if they are enabled) should catch anyway, so capping `device_api_version` like that is, as both of you already pointed out, unnecessary.

Reply:
Yup indeed, thats really the only thing that `VkApplicationInfo::apiVersion` does. Throws up validation errors if you attempt to use features of newer Vulkan versions.
In all likelihood, provided the feature is enabled, it would still work.

Reply:
Do we have a comment somewhere saying all of this you discovered now, because in a couple of months someone will likely be retreading the same steps?